### PR TITLE
BUG: Index constructor support tupleization for mixed levels

### LINKED ISSUE #18505: Creating series with multiindex from dict with tuple keys with None
#### Code Sample

```python
import pandas as pd

data1={(1, 2): 3,
       (4, None): 6}

s1 = pd.Series(data=data1)

data2={(1, 2): 3,
       (None, 5): 6}

s2 = pd.Series(data=data2)
```

#### Problem description

`s1` has MultiIndex while `s2` has Index of tuple just because the second element of the `data2` has `None` as the first element of the key tuple.

```python
>>> s1
1  2      3.0
4  NaN    NaN
dtype: float64
>>> s2
(1, 2)       3
(None, 5)    6
dtype: int64
```
Is this a bug? or is this how it's supposed to be?

FYI, both of `mi1` and `mi2` below yield MultiIndex as I expect.

```python
mi1 = pd.MultiIndex.from_tuples([(1, 2), (4, None)])
mi2 = pd.MultiIndex.from_tuples([(1, 2), (None, 5)])
```

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #18505
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.22.0.txt [hunk 1]
```diff
@@ -147,6 +147,7 @@ Indexing
 - Bug in :func:`DataFrame.groupby` where tuples were interpreted as lists of keys rather than as keys (:issue:`17979`, :issue:`18249`)
 - Bug in :func:`MultiIndex.remove_unused_levels`` which would fill nan values (:issue:`18417`)
 - Bug in :func:`MultiIndex.from_tuples`` which would fail to take zipped tuples in python3 (:issue:`18434`)
+- Bug in :class:`Index`` initialization from list of mixed type tuples (:issue:`18505`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: initialization -> construction

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -106,6 +106,16 @@ def test_construction_list_mixed_tuples(self):
         assert isinstance(idx2, Index)
         assert not isinstance(idx2, MultiIndex)
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: make this is a new test.

parametrize on input types (iterator, tuple, list) & (np.nan, None) for a missing value

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some comments. reducing code is always +1 from me!

File: pandas/tests/indexes/test_base.py (hunk 2 above)
Comment (toobaz) [reply pandas/tests/indexes/test_base.py]: (done)

@jreback ping

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -106,6 +106,16 @@ def test_construction_list_mixed_tuples(self):
...
 
+    @pytest.mark.parametrize('na_value', [None, np.nan])
+    @pytest.mark.parametrize('vtype', [list,
+                                       tuple,
+                                       lambda l: (x for x in l)])
+    def test_construction_list_tuples_nan(self, na_value, vtype):
+        # GH 18505 : valid tuples containing NaN
+        values = [(1, 'two'), (3., na_value)]
+        idx = Index(vtype(values))
+        assert isinstance(idx, MultiIndex)
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: can you construct the actual index and use ``assert_index_equal`` here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by toobaz** [comment]
@jreback ping

**Comment by jorisvandenbossche** [comment]
Do you know why the try except is no longer needed? Or was it not covered anyhow?

**Comment by toobaz** [comment]
> Do you know why the try except is no longer needed?

Not really... apparently, at some moment mixed levels weren't supported.

**Comment by jorisvandenbossche** [comment]
And codecov also indicates the tests never reached the except block, so probably indeed fixed in `from_tuples`

**Comment by jorisvandenbossche** [comment]
Can you add a test case for the original series issue as well?

**Comment by toobaz** [comment]
> Can you add a test case for the original series issue as well?

Done... but it xfails due to #18480

**Comment by jorisvandenbossche** [comment]
For me it doesn't fail, as it is None, not nan

**Comment by jorisvandenbossche** [comment]
Sorry, ignore the previous, missed the NaN in the output

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**